src/IECoreScene/ShaderNetworkAlgo.cpp

```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+			if( !boost::regex_match( destName.c_str(), splineElementMatch, g_splineElementRegex) )
+			{
+				continue;
+			}
+
+			IECore::InternedString parameterName( &destName[splineElementMatch.position(1)], splineElementMatch.length(1) );
+			auto findParameter = origParameters.find( parameterName );
+			if( findParameter == origParameters.end() )
+			{
+				continue;
+			}
+
+			const SplinefColor3fData* colorSplineData = runTimeCast<const SplinefColor3fData>( findParameter->second.get() );
+			const SplineffData* floatSplineData = runTimeCast<const SplineffData>( findParameter->second.get() );
+
+			if( !( colorSplineData || floatSplineData ) )
+			{
+				continue;
+			}
+
+			InternedString adapterHandle;
+			int knotOffset;
+
+			// Insert a conversion shader to handle connection to component
+			auto inserted = currentSplineArrayAdapters.insert( { parameterName, std::make_pair( IECore::InternedString(), 0 ) } );
+			if( inserted.second )
+			{
+				IECore::InternedString splineValuesName = parameterName.string() + "Values";
+				auto findValues = newParameters.find( splineValuesName );
+				ShaderPtr adapter;
+				int adapterArraySize = 0;
+				if( colorSplineData )
+				{
+					knotOffset = getEndPointDuplication( colorSplineData->readable().basis ).first;
+					adapter = new Shader( "Utility/ColorArrayInputAdapter", "osl:shader" );
+
+					const Color3fVectorData *splineValuesData = findValues != newParameters.end() ? runTimeCast<const Color3fVectorData>( findValues->second.get() ) : nullptr;
+					if( !splineValuesData )
+					{
+						throw IECore::Exception( "Internal failure in convertToOSLConventions - expandSpline did not create values." );
+					}
+
+					const std::vector< Imath::Color3f > &splineValues = splineValuesData->readable();
+
+					adapterArraySize = splineValues.size();
+
+					if( adapterArraySize <= maxArrayInputAdapterSize )
+					{
+						for( unsigned int i = 0; i < splineValues.size(); i++ )
+						{
+							adapter->parameters()[ g_arrayInputNames[i] ] = new Color3fData( splineValues[i] );
+						}
+					}
+				}
+				else
+				{
+					knotOffset = getEndPointDuplication( floatSplineData->readable().basis ).first;
+					adapter = new Shader( "Utility/FloatArrayInputAdapter", "osl:shader" );
+
+					const FloatVectorData *splineValuesData = findValues != newParameters.end() ? runTimeCast<const FloatVectorData>( findValues->second.get() ) : nullptr;
+					if( !splineValuesData )
+					{
+						throw IECore::Exception( "Internal failure in convertToOSLConventions - expandSpline did not create values" );
+					}
+
+					const std::vector< float > &splineValues = splineValuesData->readable();
+
+					adapterArraySize = splineValues.size();
+
+					if( adapterArraySize <= maxArrayInputAdapterSize )
+					{
+						for( unsigned int i = 0; i < splineValues.size(); i++ )
+						{
+							adapter->parameters()[ g_arrayInputNames[i] ] = new FloatData( splineValues[i] );
+						}
+					}
+				}
```

Comment:
This is just a couple of the branch above, but with different types, right? I think there's enough code here that it's probably worth templating out into a separate function. Long functions have their place, but this one has so much nested control flow that I think it could be a fair bit clearer if it were split into separate parts.